blog/2023-08-01-tolgee-with-next-13.md

```diff
@@ -0,0 +1,347 @@
+
+Now, let's establish a shared configuration that will apply to both the client and server.
+
+> For this to work, create a project in tolgee platform, get the api key in integration section. Also I assume you have your exported language files in `i18n` folder (as is visible in the file structure above).
+
+```tsx
+// shared.ts
+
+import { DevTools, Tolgee, FormatSimple } from '@tolgee/web';
+
+export const ALL_LOCALES = ['en', 'de'];
+
+export const DEFAULT_LOCALE = 'en';
+
+const apiKey = process.env.NEXT_PUBLIC_TOLGEE_API_KEY;
+const apiUrl = process.env.NEXT_PUBLIC_TOLGEE_API_URL;
+
+export async function getStaticData(languages: string[]) {
+  const result: Record<string, any> = {};
+  for (const lang of languages) {
+    result[lang] = (await import(`../i18n/${lang}.json`)).default;
+  }
+  return result;
+}
+
+export function TolgeeBase() {
+  return (
+    Tolgee()
+      .use(FormatSimple())
+      .use(DevTools())
+      // Preset shared settings
+      .updateDefaults({
+        apiKey,
+        apiUrl,
+      })
+  );
+}
+```
+
+The client part remains largely unchanged from the `pages` directory setup. It serves the purpose of translating client components and also enables in-context functionality for server-rendered components.
+
+```tsx
+// client.tsx
+
+'use client';
+
+import { TolgeeBase } from './shared';
+import { TolgeeProvider, useTolgeeSSR } from '@tolgee/react';
+import { useRouter } from 'next/navigation';
+import { useEffect } from 'react';
+
+type Props = {
+  locales: any;
+  locale: string;
+  children: React.ReactNode;
+};
+
+const tolgee = TolgeeBase().init();
+
+export const TolgeeNextProvider = ({ locale, locales, children }: Props) => {
+  // This synchronizes Tolgee for the server and client's initial render,
+  // ensuring proper initialization with language and cache  const tolgeeSSR = useTolgeeSSR(tolgee, locale, locales);
+  const router = useRouter();
+
+  useEffect(() => {
+    const { unsubscribe } = tolgeeSSR.on('permanentChange', () => {
+      // This ensures that server components refresh after
+      // translation modifications using in-context
+      router.refresh();
+    });
+
+    return () => unsubscribe();
+  }, [tolgeeSSR, router]);
+
+  return (
+    <TolgeeProvider tolgee={tolgeeSSR} options={{ useSuspense: false }}>
+      {children}
+    </TolgeeProvider>
+  );
+};
+```
+
+The only significant change is the listener for the `permanentChange` event. This event triggers when a translation is updated through an in-context dialog, allowing for a server component refresh with `router.refresh`.
+
+Now, let's proceed to the server part. As server components don't support React hooks, we need to recreate similar abstractions to those found in `@tolgee/react` ourselves. Fortunately, this process is relatively straightforward; we can essentially utilize vanilla Tolgee, with proper configuration and caching of the instance.
```

Comment:
So we want our users to recreate the abstractions? Isn't that why the users use libraries like ours?